### ISSUE TITLE
`TabClosed` callbacks break if `expand("<afile>")` is called prior

### ISSUE THREAD

**Issue by NGPONG** [sindrets/diffview.nvim#369]
Is this an expected behavior or a bug? Because after exploring a little deeper, I found that closing multiple panels is done via `TabClosed` event callbacks, defined here

https://github.com/sindrets/diffview.nvim/blob/86bf6182e2ea767c21711de8a3c396d9b635d970/lua/diffview/init.lua#L44-L50

However, the 'TabClosed' event is specified to be non-recursive (event cannot trigger itself) (I didn't find this in the `:h tabclosed`, but I found this logic in the neovim source code [static void do_autocmd_winclosed(win_T *win)](https://github.com/neovim/neovim/blob/fdc8e966a9183c08f2afec0817d03b7417a883b3/src/nvim/window.c#L3018C45-L3030)), so the end result is that the TabClosed event will not be triggered again when the close operation is performed a second time and the remaining panels will not be closed further down.

Now, I got around this by forcing the event to fire again by calling vim.api.nvim_exec_autocmds in the TabClosed callback, but I'm not sure if that's the right solution, or if the result (only two panels closed) is what you expect.

A small addition, if this is a bug, can you try to let me fix it? I really like this plugin and would like to give some modest contribution.

**Comment by sindrets** [comment]
Just because this is terminology that often confuses users of vim, and because I'm confused by the issue you're describing; I'm gonna include the summary from `:h windows-intro`:

```
Summary:
   A buffer is the in-memory text of a file.
   A window is a viewport on a buffer.
   A tab page is a collection of windows.
```

Please clarify: are you talking about vim tabs (the visible collection of windows), or are you talking about vim windows?

> DiffviewClose or tabclose will only close a maximum of two tabs

No, `:tabclose` will only ever close *one* tab page. You can specify which one with `:tabclose [count]`.

If you're actually talking about windows, and you're saying that `:tabclose` doesn't close all the windows in the current tab, then that would be unusual. In this case I would ask that you try to reproduce with a minimal config. You can [find a template here](https://gist.github.com/sindrets/3e671e79ce4566667f9ec176786dfadd).

> A small addition, if this is a bug, can you try to let me fix it? I really like this plugin and would like to give some modest contribution.

I really appreciate the incentive!

**Comment by NGPONG** [comment]
https://github.com/vim/vim/blob/59f7038536a370d771758dc34036cc1424be7421/src/autocmd.c#L1984

https://github.com/vim/vim/blob/8060687905bdadc46abb68ee6d40e5660e352297/src/ex_docmd.c#L10652-L10659

https://github.com/vim/vim/blob/8060687905bdadc46abb68ee6d40e5660e352297/src/ex_docmd.c#L10623-L10640

The initialization logic for '<afile>'(autocmd_fname) and '<amatch>'(autocmd_match) is found in the three places I mentioned above(here). I didn't have enough time to preview it carefully, but my intuition tells me that the processing logic of '<afile>' and '<amatch>' is extremely similar in some cases. So I think there may be some unknown risks in the temporary replacement of '<amatch>'.

> If you wish to have a go at it, please go ahead.

One of the immediate priorities is to issue a new issuse for upstream and continue to explore this issue(neovim/neovim#23941). I wouldn't be reckless enough to add unknown risks to your project, and I'm not sure that this solution will destroy the ideas and programming you have in this project. So before I start I would like to ask your opinion again, which I respect very much. @sindrets

**Comment by NGPONG** [comment]
Hmm, it looks like this bug has been around for a while. But thankfully it's making some headway now. neovim/neovim#18964

**Comment by sindrets** [comment]
Very nice, it's now been fixed upstream. However, while I see that the fix is scheduled for backporting to v0.9, we aim to support 2 versions behind the current stable release. I suggest we go through with implementing the workaround of using `match` instead when `vim.fn.has("nvim-0.9.2") ~= 1`.

**Comment by NGPONG** [comment]
> I suggest we go through with implementing the workaround of using match instead when vim.fn.has("nvim-0.9.2") ~= 1.

Wise move! I'm working on it.
